src/core/tools/common/asio_net_provider.cpp

```diff
@@ -102,7 +102,8 @@ namespace dsn {
         rpc_session_ptr asio_network_provider::create_client_session(::dsn::rpc_address server_addr)
         {
             auto sock = std::shared_ptr<boost::asio::ip::tcp::socket>(new boost::asio::ip::tcp::socket(_io_service));
-            return rpc_session_ptr(new asio_rpc_session(*this, server_addr, sock, new_message_parser(), true));
+            auto parser = new_message_parser(_client_hdr_format);
+            return rpc_session_ptr(new asio_rpc_session(*this, server_addr, sock, parser, true));
```

Comment:
The change from unique_ptr to ref_counter introduces performance overhead. I see it seems because you want to put it inside message as well, but these two together are costly. Any better approaches?